docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,264 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+=========================
+The Arrow C data protocol
+=========================
+
+*DRAFT*
+
+Data type description - format strings
+======================================
+
+TODO:
+
+Structure definitions
+=====================
+
+.. code-block:: c
+
+   struct ArrowArray {
+     // Type description
+     const char* format;
+     const char* metadata;
+
+     // Data description
+     int64_t length;
+     int64_t null_count;
+     int64_t offset;
+     int64_t n_buffers;
+     int64_t n_children;
+     const void **buffers;
+     struct ArrowArray** children;
+     struct ArrowArray* dictionary;
+
+     // Release callback
+     void (*release)(struct ArrowArray*);
+     // Opaque producer-specific data
+     void* private_data;
+   };
+
+The ArrowArray structure
+------------------------
+
+Members
+'''''''
+
+.. c:member:: const char* ArrowArray.format
+
+   Mandatory.  A null-terminated, ASCII-encoded string describing
+   the data type.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   Consumers MAY decide not to support all data types, but they
+   should document this limitation.
+
+.. c:member:: const char* ArrowArray.metadata
+
+   Optional.  A null-terminated, ASCII-encoded string describing
```

Comment:
I think this is inconsistent with flatbuffers which supports UTF-8,

Reply:
Perhaps. I don't have a strong feeling here. I assumed that full ASCII would be easier to support (at the cost of a limitation for users), but perhaps UTF8 is reasonable in all languages nowadays :-)